### PR TITLE
Use ophyd-async DirectoryProvider

### PR DESCRIPTION
**Pull request by DiamondJoseph**
Exposes a global DirectoryProvider for ophyd-async devices to make use of, with a setter for services making use of dodal (e.g. BlueAPI) to choose their own implementation. Solves #305 
Extracted from https://github.com/DiamondLightSource/dodal/tree/directory_provider to allow i22 changes to be isolated and 

### Instructions to reviewer on how to test:
1. Instantiate an Ophyd-async device which requires a DirectoryProvider in its init
2. Confirm that the device can write to the directory provided.

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)

### CODE REVIEW

**Review by DominicOram — APPROVED** [review]
This looks good to me, thanks. Do we have some docs on the `DirectoryProvider` and how it should be used?

**Comment by DiamondJoseph** [comment]
> Do we have some docs on the `DirectoryProvider` and how it should be used?

[There's some incidental in this PR of documentation to ophyd-async](https://github.com/bluesky/ophyd-async/pull/107), which I will try and get over the line along with all the outstanding changes in dodal.

TLDR: call it, get a directory and an optional prefix for the files that an ophyd-async device is going to write. BlueAPI will have an implementation which calls off to GDA to get the current visit directory and current scan number and use those.
